### PR TITLE
[MI-2262][server]: Created plugin API to request branches for a repo.

### PR DESCRIPTION
**Pull request by avas27JTG**
Created plugin API to request branches for a repo.

### CODE REVIEW

File: server/plugin/api.go [hunk 1]
```diff
@@ -749,6 +750,35 @@ func (p *Plugin) handleGetGitRepositories(w http.ResponseWriter, r *http.Request
...
+	if err != nil {
+		p.API.LogError("Error in fetching git repository branches", err.Error())
+		p.handleError(w, r, &serializers.Error{Code: statusCode, Message: err.Error()})
+		return
+	}
+
+	// Azure DevOps return branch name as "refs/heads/<branch-name>", but we need to use only "<branch-name>" so, remove unused part from the name
+	for index, value := range response.Value {
+		response.Value[index].Name = value.Name[11:]
+	}
```
Comment (ayusht2810) [inline server/plugin/api.go:777]:
```suggestion
	for _, value := range response.Value {
		value.Name = value.Name[11:]
	}
```

Comment (ayusht2810) [reply server/plugin/api.go:777]: check if this works
also, are we always sure that this length is greater than 11?

Comment (avas27JTG) [reply server/plugin/api.go]: yes it will be always >= 11, still added the checks

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/plugin/api.go [hunk 2]
```diff
@@ -749,6 +750,37 @@ func (p *Plugin) handleGetGitRepositories(w http.ResponseWriter, r *http.Request
 	p.writeJSON(w, response.Value)
 }
 
+func (p *Plugin) handleGetGitRepositoryBranches(w http.ResponseWriter, r *http.Request) {
+	mattermostUserID := r.Header.Get(constants.HeaderMattermostUserID)
+
+	pathParams := mux.Vars(r)
+	organization := pathParams[constants.PathParamOrganization]
+	project := pathParams[constants.PathParamProject]
+	repository := pathParams[constants.PathParamRepository]
```
Comment (manojmalik20) [inline server/plugin/api.go:759]:
```suggestion
	organization := strings.TrimSpace(pathParams[constants.PathParamOrganization])
	project := strings.TrimSpace(pathParams[constants.PathParamProject])
	repository := strings.TrimSpace(pathParams[constants.PathParamRepository])
```

File: server/plugin/api.go [hunk 3]
```diff
@@ -749,6 +750,37 @@ func (p *Plugin) handleGetGitRepositories(w http.ResponseWriter, r *http.Request
...
+	}
+
+	response, statusCode, err := p.Client.GetGitRepositoryBranches(organization, project, repository, mattermostUserID)
+	if err != nil {
+		p.API.LogError("Error in fetching git repository branches", err.Error())
+		p.handleError(w, r, &serializers.Error{Code: statusCode, Message: err.Error()})
+		return
+	}
+
+	// Azure DevOps return branch name as "refs/heads/<branch-name>", but we need to use only "<branch-name>" so, remove unused part from the name
```
Comment (manojmalik20) [inline server/plugin/api.go:774]:
```suggestion
	// Azure DevOps returns branch name as "refs/heads/<branch-name>", but we need to use only "<branch-name>" so, remove unused part from the name
```

File: server/plugin/api_test.go [hunk 4]
```diff
@@ -975,3 +975,67 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
 }
+
+func TestHandleGetGitRepositoryBranches(t *testing.T) {
+	defer monkey.UnpatchAll()
+	p := Plugin{}
+	mockAPI := &plugintest.API{}
+	mockCtrl := gomock.NewController(t)
+	mockedClient := mocks.NewMockClient(mockCtrl)
+	p.API = mockAPI
+	p.Client = mockedClient
```
Comment (manojmalik20) [inline server/plugin/api_test.go:986]: Make a function for this and use that everywhere.

File: server/plugin/api_test.go [hunk 5]
```diff
@@ -975,3 +975,67 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+	for _, testCase := range []struct {
+		description                 string
+		Organization                string
+		Repository                  string
+		Project                     string
+		GetGitRepositoryBranchesErr error
+		statusCode                  int
```
Comment (manojmalik20) [inline server/plugin/api_test.go:993]: Make casing of the variables consistent.

File: server/plugin/api_test.go [hunk 6]
```diff
@@ -975,3 +975,67 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+			description:                 "HandleGetGitRepositoryBranches: GetGitRepositoryBranches returns error",
+			Organization:                "mockOrganization",
+			Project:                     "mockProject",
+			Repository:                  "mockRepository",
+			GetGitRepositoryBranchesErr: errors.New("failed to get git repository branches"),
+			statusCode:                  http.StatusInternalServerError,
+		},
+	} {
+		t.Run(testCase.description, func(t *testing.T) {
+			mockAPI.On("LogError", mock.AnythingOfType("string"), mock.AnythingOfType("string"), mock.AnythingOfType("string"))
```
Comment (manojmalik20) [inline server/plugin/api_test.go:1018]: Use util function here.

File: server/plugin/api_test.go [hunk 7]
```diff
@@ -975,3 +975,67 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+
+			if testCase.statusCode == http.StatusInternalServerError || testCase.statusCode == http.StatusOK {
+				mockedClient.EXPECT().GetGitRepositoryBranches(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(&serializers.GitBranchesResponse{}, testCase.statusCode, testCase.GetGitRepositoryBranchesErr)
+			}
+
+			req := httptest.NewRequest(http.MethodGet, "/branches", bytes.NewBufferString(`{}`))
```
Comment (manojmalik20) [inline server/plugin/api_test.go:1024]: Use constant for the path.

File: server/plugin/api_test.go [hunk 8]
```diff
@@ -975,3 +975,67 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
+				"project":      testCase.Project,
+				"repository":   testCase.Repository,
+			}
+
+			req = mux.SetURLVars(req, pathParams)
+
+			w := httptest.NewRecorder()
+			p.handleGetGitRepositoryBranches(w, req)
+			resp := w.Result()
+			assert.Equal(t, testCase.statusCode, resp.StatusCode)
```
Comment (manojmalik20) [inline server/plugin/api_test.go:1038]: You should also check the response along with the status code.

File: server/plugin/client_test.go [hunk 9]
```diff
@@ -336,3 +336,44 @@ func TestCall(t *testing.T) {
...
 	}
 }
+
+func TestGetGitRepositoryBranches(t *testing.T) {
+	defer monkey.UnpatchAll()
+	p := Plugin{}
+	mockAPI := &plugintest.API{}
+	p.API = mockAPI
+	c := InitClient(&p)
+	p.Client = c
```
Comment (manojmalik20) [inline server/plugin/client_test.go:346]: Use function for this.

File: server/plugin/api_test.go [hunk 10]
```diff
@@ -977,58 +963,140 @@ func TestGetUserChannelsForTeam(t *testing.T) {
...
 		{
 			description:                 "HandleGetGitRepositoryBranches: GetGitRepositoryBranches returns error",
-			Organization:                "mockOrganization",
-			Project:                     "mockProject",
-			Repository:                  "mockRepository",
-			GetGitRepositoryBranchesErr: errors.New("failed to get git repository branches"),
+			organization:                "mockOrganization",
+			project:                     "mockProject",
+			repository:                  "mockRepository",
+			getGitRepositoryBranchesErr: errors.New("failed to get git repository branches"),
```
Comment (manojmalik20) [inline server/plugin/api_test.go:1081]: Use constant here also.

Comment (raghavaggarwal2308) [reply server/plugin/api_test.go:1081]: no constant created for this

Comment (raghavaggarwal2308) [reply server/plugin/api_test.go:1081]: I can create a constant in the testcase itself

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**